atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/OracleTableNameUnmapper.java

```diff
@@ -15,34 +15,75 @@
  */
 package com.palantir.atlasdb.keyvalue.dbkvs;
 
+import java.sql.SQLException;
+import java.util.concurrent.ExecutionException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
 import com.google.common.collect.Iterables;
 import com.palantir.atlasdb.AtlasDbConstants;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.dbkvs.impl.ConnectionSupplier;
 import com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs;
 import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
 import com.palantir.nexus.db.sql.AgnosticResultSet;
+import com.palantir.nexus.db.sql.SqlConnection;
 
 class OracleTableNameUnmapper {
+    private static final Logger log = LoggerFactory.getLogger(OracleTableNameUnmapper.class);
+
     private final ConnectionSupplier conns;
 
+    private LoadingCache<String, String> unmappingCache = CacheBuilder.newBuilder().build(
+            new CacheLoader<String, String>() {
+                @Override
+                public String load(String fullTableName) throws Exception {
+                    SqlConnection conn = null;
+                    try {
+                        conn = conns.getFresh();
```

Comment:
This will block on getting a new connection which might lead to long pauses/deadlocks, should we switch back to reusing a connection from the supplier?